### shapes PyImage:-https://pyimagesearch.com/2016/02/08/opencv-shape-detection/?_ga=2.230735647.1271202568.1701200962-1743169504.1688156425#
### contours PyImage:-https://pyimagesearch.com/2016/02/01/opencv-center-of-contour/
### contour aprrzomation:-https://pyimagesearch.com/2021/10/06/opencv-contour-approximation/

In [1]:
import cv2
import pandas as pd
import sklearn as sk
import numpy as np
import matplotlib.pyplot as plt

In [2]:
img1 = cv2.imread(r"E:\PyImage_ComputerVision\ObjectDetection\WorkingData\InputData\shape1.png")
img2 = cv2.imread(r"E:\PyImage_ComputerVision\ObjectDetection\WorkingData\InputData\shape2.png")

In [3]:
img1.shape

(561, 600, 3)

In [4]:
img2.shape

(400, 500, 3)

In [7]:
gray1 = cv2.cvtColor(img1,cv2.COLOR_BGR2GRAY)
gray2 = cv2.cvtColor(img2,cv2.COLOR_BGR2GRAY)

In [8]:
cv2.imshow("Gray Image 1",gray1)
cv2.imshow("Gray Image 2",gray2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [9]:
thresh1 = cv2.threshold(gray1,0,255,cv2.THRESH_BINARY | cv2.THRESH_OTSU)
thresh2 = cv2.threshold(gray2,0,255,cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)

In [13]:
print(thresh1[0])
print(thresh2[0])

98.0
165.0


In [14]:
cv2.imshow("Threshold Image 1",thresh1[1])
cv2.imshow("Threshold Image 2",thresh2[1])
cv2.waitKey(0)
cv2.destroyAllWindows()

# --------------------------------------------------------------------------------------------

## cv2.findContours() Method:-https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#gadf1ad6a0b82947fa1fe3c3d497f260e0

# Flags:-

## retirval mode flags:-https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#ga819779b9857cc2f8601e6526a3a5bc71
## Contour Approxmimation Flags:-https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#ga4303f45752694956374734a03c54d5ff

In [15]:
c1 = cv2.findContours(thresh1[1],cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

In [17]:
type(c1)

tuple

In [18]:
len(c1)

2

## It returns a atuple contianing (contours) and (Hierachy)

In [20]:
fg1 = c1[0]
fg1

(array([[[243, 397]],
 
        [[242, 398]],
 
        [[224, 398]],
 
        [[224, 400]],
 
        [[223, 401]],
 
        [[223, 402]],
 
        [[222, 403]],
 
        [[222, 404]],
 
        [[221, 405]],
 
        [[221, 407]],
 
        [[220, 408]],
 
        [[220, 409]],
 
        [[219, 410]],
 
        [[219, 411]],
 
        [[218, 412]],
 
        [[218, 413]],
 
        [[217, 414]],
 
        [[217, 416]],
 
        [[216, 417]],
 
        [[216, 418]],
 
        [[215, 419]],
 
        [[215, 421]],
 
        [[214, 422]],
 
        [[214, 423]],
 
        [[213, 424]],
 
        [[213, 425]],
 
        [[212, 426]],
 
        [[212, 427]],
 
        [[211, 428]],
 
        [[211, 429]],
 
        [[210, 430]],
 
        [[214, 434]],
 
        [[215, 434]],
 
        [[218, 437]],
 
        [[219, 437]],
 
        [[221, 439]],
 
        [[222, 439]],
 
        [[224, 441]],
 
        [[225, 441]],
 
        [[228, 444]],
 
        [[229, 444]],
 
        [[231, 4

In [21]:
type(fg1)

tuple

In [22]:
len(fg1)

18

## This means there are 18 contours detected in the Image

In [55]:
fg2 = c1[1]
fg2

array([[[ 1, -1, -1, -1],
        [ 2,  0, -1, -1],
        [ 3,  1, -1, -1],
        [ 4,  2, -1, -1],
        [ 5,  3, -1, -1],
        [ 6,  4, -1, -1],
        [ 7,  5, -1, -1],
        [ 8,  6, -1, -1],
        [ 9,  7, -1, -1],
        [10,  8, -1, -1],
        [11,  9, -1, -1],
        [12, 10, -1, -1],
        [13, 11, -1, -1],
        [14, 12, -1, -1],
        [15, 13, -1, -1],
        [16, 14, -1, -1],
        [17, 15, -1, -1],
        [-1, 16, -1, -1]]], dtype=int32)

In [24]:
type(fg2)

numpy.ndarray

In [25]:
fg2.shape

(1, 18, 4)

In [26]:
fg2[0]

array([[ 1, -1, -1, -1],
       [ 2,  0, -1, -1],
       [ 3,  1, -1, -1],
       [ 4,  2, -1, -1],
       [ 5,  3, -1, -1],
       [ 6,  4, -1, -1],
       [ 7,  5, -1, -1],
       [ 8,  6, -1, -1],
       [ 9,  7, -1, -1],
       [10,  8, -1, -1],
       [11,  9, -1, -1],
       [12, 10, -1, -1],
       [13, 11, -1, -1],
       [14, 12, -1, -1],
       [15, 13, -1, -1],
       [16, 14, -1, -1],
       [17, 15, -1, -1],
       [-1, 16, -1, -1]], dtype=int32)

## To Understand Hierarchy Array we can Refer :- https://docs.opencv.org/4.x/d9/d8b/tutorial_py_contours_hierarchy.html 

In [61]:
fg1[0]

array([[[243, 397]],

       [[242, 398]],

       [[224, 398]],

       [[224, 400]],

       [[223, 401]],

       [[223, 402]],

       [[222, 403]],

       [[222, 404]],

       [[221, 405]],

       [[221, 407]],

       [[220, 408]],

       [[220, 409]],

       [[219, 410]],

       [[219, 411]],

       [[218, 412]],

       [[218, 413]],

       [[217, 414]],

       [[217, 416]],

       [[216, 417]],

       [[216, 418]],

       [[215, 419]],

       [[215, 421]],

       [[214, 422]],

       [[214, 423]],

       [[213, 424]],

       [[213, 425]],

       [[212, 426]],

       [[212, 427]],

       [[211, 428]],

       [[211, 429]],

       [[210, 430]],

       [[214, 434]],

       [[215, 434]],

       [[218, 437]],

       [[219, 437]],

       [[221, 439]],

       [[222, 439]],

       [[224, 441]],

       [[225, 441]],

       [[228, 444]],

       [[229, 444]],

       [[231, 446]],

       [[232, 446]],

       [[235, 449]],

       [[236, 449]],

       [[2

In [60]:
fg1[0].shape

(95, 1, 2)

In [58]:
fg1[0][0]

array([[243, 397]], dtype=int32)

## Now we can see that at each contour has N number of points each point is represtend by an (1,2) numpy array :-
[[2,3]]
means x-coordinate is 2
and  y-coordinate is 3

# --------------------------------------------------------------------------------------------

In [63]:
imgfg1 = img1.copy()

## cv2.drawContours() Method:-https://docs.opencv.org/4.x/d6/d6e/group__imgproc__draw.html#ga746c0625f1781f1ffc9056259103edbc

In [64]:
cv2.drawContours(imgfg1,c1[0],-1,(0,255,0),3)
cv2.imshow("Contours Image 1",imgfg1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [66]:
z1 = img1.copy()
cv2.drawContours(z1,c1[0],1,(0,255,0),3)
cv2.imshow("Contours Image 1",z1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [67]:
z1 = img1.copy()
cv2.drawContours(z1,c1[0],3,(0,255,0),3)
cv2.imshow("Contours Image 1",z1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [69]:
z1 = img1.copy()
cv2.drawContours(z1,c1[0],19,(0,255,0),3)
cv2.imshow("Contours Image 1",z1)
cv2.waitKey(0)
cv2.destroyAllWindows()

error: OpenCV(4.8.1) D:\a\opencv-python\opencv-python\opencv\modules\imgproc\src\drawing.cpp:2526: error: (-215:Assertion failed) 0 <= contourIdx && contourIdx < (int)last in function 'cv::drawContours'


In [72]:
z1 = img1.copy()
cv2.drawContours(z1,c1[0],0,(0,255,0),3)
cv2.imshow("Contours Image 1",z1)
cv2.waitKey(0)
cv2.destroyAllWindows()

## Above we Changed Contouridx and Noticed the Results

In [74]:
z1 = img1.copy()
cv2.drawContours(z1,c1[0],-1,(0,255,0),-1)
cv2.imshow("Contours Image 1",z1)
cv2.waitKey(0)
cv2.destroyAllWindows()

## Above we changed the Thickess to -1 and saw that contours were filled

## We can also use HIerarachy array and pass it to drawContours() to change some Drawing of Countours.

# --------------------------------------------------------------------------------------------

In [39]:
maxc1 = max(c1[0],key=cv2.contourArea)

In [41]:
maxc1.shape

(322, 1, 2)

In [85]:
cnt1 = c1[0][1]
cnt1.shape

(322, 1, 2)

In [45]:
imgfg2 = img1.copy()

In [46]:
cv2.drawContours(imgfg2,maxc1,-1,(0,255,0),3)
cv2.imshow("Max Contour Image",imgfg2)
cv2.waitKey(0)
cv2.destroyAllWindows()

## We can see in the Above we are using the biggest Contour of the all and we VISUALIZED it and we can see in the Original Countour has 322 Points

## TO Calculate the Perimeter of our Contour we use `cv2.arcLength()` Method.
## If the Contour is CLosed then Perimeter is Calculated.
## If the Contour is Open then Curve Length is Calculated

## cv2.arcLenght() Method:-https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#ga8d26483c636be6b35c3ec6335798a47c

## The perimeter is the distance around the outside of a shape. It is calculated by adding up the lengths of all the sides of the shape. In OpenCV, the perimeter of a contour is calculated by adding up the lengths of all the edges of the contour.
The units of the perimeter returned by the arcLength() function depend on the units of the image that the contour was extracted from. For example, if the image is in pixels, then the perimeter will be in pixels. If the image is in millimeters, then the perimeter will be in millimeters.

In [86]:
per1 = cv2.arcLength(cnt1,True)

In [87]:
per1

454.7422024011612

In [89]:
per2 = cv2.arcLength(cnt1,False)

In [91]:
per2

453.7422024011612

## The main Functioon used for Contour Approxmiation we use `cv2.approxPolyDP()` Method

## cv2.approxPolyDP() Method:-https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#ga0012a5fdaea70b8a9970165d98722b4c

In [113]:
approx = cv2.approxPolyDP(cnt1,0.02*per1,True)

In [114]:
approx.shape

(4, 1, 2)

In [115]:
imgfg3 = img1.copy()
cv2.drawContours(imgfg3,approx,-1,(0,255,0),3)

array([[[21, 24, 23],
        [22, 25, 23],
        [23, 26, 24],
        ...,
        [22, 24, 22],
        [21, 23, 22],
        [20, 24, 22]],

       [[23, 25, 25],
        [20, 22, 20],
        [24, 24, 22],
        ...,
        [22, 23, 21],
        [22, 24, 21],
        [22, 26, 24]],

       [[22, 23, 22],
        [26, 28, 25],
        [26, 29, 26],
        ...,
        [24, 25, 24],
        [23, 24, 23],
        [20, 21, 21]],

       ...,

       [[55, 57, 58],
        [54, 55, 56],
        [53, 55, 55],
        ...,
        [40, 39, 40],
        [38, 38, 38],
        [38, 40, 40]],

       [[54, 55, 56],
        [52, 54, 53],
        [54, 56, 56],
        ...,
        [40, 41, 41],
        [39, 39, 39],
        [40, 42, 39]],

       [[58, 59, 57],
        [50, 52, 51],
        [51, 53, 53],
        ...,
        [46, 48, 48],
        [43, 43, 43],
        [41, 41, 40]]], dtype=uint8)

In [116]:
cv2.imshow("Before Approx",imgfg2)
cv2.imshow("After Approx",imgfg3)
cv2.waitKey(0)
cv2.destroyAllWindows()

# --------------------------------------------------------------------------------------------

In [118]:
cnt2 = cv2.findContours(thresh2[1],cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

In [129]:
len(cnt2[0])

1

In [130]:
mc2 = cnt2[0][0]
mc2

array([[[ 99, 106]],

       [[ 99, 128]],

       [[102, 131]],

       [[103, 131]],

       [[106, 134]],

       [[107, 134]],

       [[110, 137]],

       [[111, 137]],

       [[113, 139]],

       [[114, 139]],

       [[117, 142]],

       [[118, 142]],

       [[123, 147]],

       [[124, 147]],

       [[136, 159]],

       [[136, 160]],

       [[142, 166]],

       [[143, 166]],

       [[154, 177]],

       [[154, 183]],

       [[150, 187]],

       [[145, 187]],

       [[139, 181]],

       [[138, 181]],

       [[121, 164]],

       [[121, 163]],

       [[117, 159]],

       [[116, 159]],

       [[111, 154]],

       [[110, 154]],

       [[107, 151]],

       [[106, 151]],

       [[103, 148]],

       [[102, 148]],

       [[ 99, 145]],

       [[ 99, 250]],

       [[109, 250]],

       [[110, 249]],

       [[125, 249]],

       [[126, 248]],

       [[136, 248]],

       [[137, 247]],

       [[143, 247]],

       [[144, 246]],

       [[146, 246]],

       [[1

## There is one Contour in the Image 

In [151]:
per2 = cv2.arcLength(mc2,True)

In [152]:
per2

1467.8792823553085

In [153]:
apx2 = cv2.approxPolyDP(mc2,per2*0.05,True)

In [154]:
mc2.shape

(236, 1, 2)

In [155]:
apx2.shape

(4, 1, 2)

In [157]:
out1 = img2.copy()
out2 = img2.copy()
cv2.drawContours(out1,[mc2],-1,(0,255,0),2)
cv2.drawContours(out2,[apx2],-1,(0,255,0),2)
cv2.imshow("Before Approx",out1)
cv2.imshow("After Approx",out2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [158]:
def DetectShape(contour):
    per = cv2.arcLength(contour,True)
    shape = None
    apx = cv2.approxPolyDP(contour,0.04 * per,True)
    vertices = apx.shape[0]
    if vertices > 4:
        if vertices == 5:
            shape = "Polygon"
        else:
            shape = "Circle"
    else:
        if vertices == 4:
            (x,y,w,h) = cv2.boundingRect(apx)
            rt = float(w)/float(h)
            shape = "Square" if rt >= 0.95 and rt <= 1.05 else "Rectangle"
        if vertices == 3:
            shape = "Traingle"
    return shape

In [190]:
cnts = cv2.findContours(thresh1[1],cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

In [191]:
shape_out = img1.copy()
for idx,cnt in enumerate(cnts[0][0:17]):
    shape = DetectShape(cnt)
    M = cv2.moments(cnt)
    print(idx)
    Cx = int(M["m10"] / M["m00"])
    Cy = int(M["m01"] / M["m00"])
    cv2.drawContours(shape_out,[cnt],-1,(0,255,0),2)
    cv2.putText(shape_out,shape,(Cx,Cy),cv2.FONT_HERSHEY_SIMPLEX,0.5,(255,255,255),2)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16


In [192]:
cv2.imshow("Detected Shapes",shape_out)
cv2.waitKey(0)
cv2.destroyAllWindows()